File: device-app/ios-swift/umsDemo/ums/Data/DataSource/MessageLocalDataSource.swift
import Foundation

class MessageLocalDataSource: JsonLocalDataSource<[DataDef.Message]> {
    private static let fileName = "messages.json"

    static var shared = MessageLocalDataSource()

    let messages: Observable<[DataDef.Message]> = Observable([])
    private var messageIdMappings = [String: DataDef.Message]()

    private let lock = NSLock()

    private init() {
        super.init(Self.fileName)
        loadRecentMessages()
    }

    func addMessage(_ item: DataDef.Message) {
        NSLog("[\(type(of: self))] addMessage(\(item.id))")

        if containsMessage(item.id) { return }

        lock.lock()
        messages.value.append(item)
        messages.value.sort { $0.sentAt > $1.sentAt }
        messageIdMappings[item.id] = item
        lock.unlock()

        messages.notifyObservers()
        save()
    }

    func containsMessage(_ messageId: String) -> Bool {
        lock.lock()
        let result = messageIdMappings.keys.contains(messageId)
        lock.unlock()

        return result
    }

    func getMessage(_ messageId: String) -> DataDef.Message? {
        lock.lock()
        let result = messageIdMappings[messageId]
        lock.unlock()

        return result
    }

    func removeMessage(_ messageId: String) {
        NSLog("[\(type(of: self))] removeMessage(\(messageId))")

        lock.lock()
        messages.value.removeAll { $0.id == messageId }
        lock.unlock()

        messages.notifyObservers()
        save()
    }

    func removeMessages(_ messageIds: Set<String>) {
        NSLog("[\(type(of: self))] removeMessage([\(messageIds.joined(separator: ", "))])")

        lock.lock()
        messages.value.removeAll { messageIds.contains($0.id) }

        for messageId in messageIds {
            messageIdMappings.removeValue(forKey: messageId)
        }
        lock.unlock()

        messages.notifyObservers()
        save()
    }

    func markAsRead(_ messageId: String) {
        NSLog("[\(type(of: self))] setAsRead(\(messageId))")

        lock.lock()
        if let message = messageIdMappings[messageId] {
            message.isRead = true
        }
        lock.unlock()

        messages.notifyObservers()
        save()
    }

    func markAllAsRead() {
        NSLog("[\(type(of: self))] setAllAsRead()")

        lock.lock()
        for message in messages.value {
            message.isRead = true
        }
        lock.unlock()

        messages.notifyObservers()
        save()
    }
    
    func updateDeliveryInfo(_ messageId: String, mAlimtalkSendTime: Int64, mMunjaSendTime: Int64, mMunjaType: Int) -> DataDef.Message? {
        NSLog("[\(type(of: self))] updateDeliveryInfo(\(messageId), mAlimtalkSendTime: \(mAlimtalkSendTime), mMunjaSendTime: \(mMunjaSendTime), mMunjaType: \(mMunjaType))")
        
        if let message = messageIdMappings[messageId] {
            if mAlimtalkSendTime > 0 {
                message.alimTalkSentAt = Date(timeIntervalSince1970: TimeInterval(mAlimtalkSendTime) / 1000)
            }
            
            if mMunjaSendTime > 0 {
                message.munjaSentAt = Date(timeIntervalSince1970: TimeInterval(mMunjaSendTime) / 1000)
            }
            
            message.munjaType = DataDef.MunjaType(rawValue: mMunjaType)
            save()
            
            return message
        }
        
        return nil
    }

    func save() {
        NSLog("[\(type(of: self))] save()")
        value = messages.value
    }

    override func delete() -> Bool {
        let result = super.delete()
        Self.shared = MessageLocalDataSource()
        return result
    }

    private func loadRecentMessages() {
        if let array: [DataDef.Message] = value {
            let now = Date()
            let retainDay: Int = Config.shared?.umsMsgRetainDay ?? 0

            for message in array {
                let days: Int = Calendar.current.dateComponents([.day], from: message.sentAt, to: now).day!

                if days < retainDay {
                    messages.value.append(message)
                    messageIdMappings[message.id] = message
                }
            }

            messages.value.sort { $0.sentAt > $1.sentAt }

            if array.count != messages.value.count {
                save()
            }
        }
    }
}

File: device-app/android/Java/umsDemo/app/src/main/java/com/rationalowl/umsdemo/data/datasource/MessageLocalDataSource.java
package com.rationalowl.umsdemo.data.datasource;

import android.util.Log;

import com.rationalowl.umsdemo.data.Config;
import com.rationalowl.umsdemo.data.DataDef;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;

public class MessageLocalDataSource extends JsonLocalDataSource<DataDef.Message[]> {
    private static final String TAG = "MessagesRepository";
    private static final String FILE_NAME = "messages.json";

    public interface OnMessagesChangedListener {
        void onMessagesChange(List<DataDef.Message> messages);
    }

    private static MessageLocalDataSource instance;

    public static synchronized MessageLocalDataSource getInstance() {
        if (instance == null) {
            instance = new MessageLocalDataSource();
        }

        return instance;
    }

    private final List<DataDef.Message> messages = new ArrayList<>();
    private final Map<String, DataDef.Message> messageIdMappings = new HashMap<>();

    private OnMessagesChangedListener onMessagesChangeListener;

    public MessageLocalDataSource() {
        super(DataDef.Message[].class, FILE_NAME);
        loadRecentMessages();
    }

    public void setOnMessagesChangeListener(OnMessagesChangedListener listener) {
        onMessagesChangeListener = listener;
    }

    public void addMessage(DataDef.Message message) {
        Log.d(TAG, "addMessage(" + message.getId() + ")");

        if (containsMessage(message.getId())) return;

        synchronized (this) {
            messages.add(message);
            messages.sort((m1, m2) -> m2.getSentAt().compareTo(m1.getSentAt()));
            messageIdMappings.put(message.getId(), message);
        }

        onMessagesChange();
        save();
    }

    public synchronized boolean containsMessage(String messageId) {
        return messageIdMappings.containsKey(messageId);
    }

    public synchronized DataDef.Message getMessage(String messageId) {
        return messageIdMappings.get(messageId);
    }

    public List<DataDef.Message> getMessages() {
        return messages;
    }

    public void removeMessage(String messageId) {
        Log.d(TAG, "removeMessage(" + messageId + ")");

        synchronized (this) {
            messages.removeIf(message -> message.getId().equals(messageId));
            messageIdMappings.remove(messageId);
        }

        onMessagesChange();
        save();
    }

    public void removeMessages(Set<String> messageIds) {
        Log.d(TAG, "removeMessages([" + String.join(", ", messageIds) + "])");

        synchronized (this) {
            messages.removeIf(message -> messageIds.contains(message.getId()));

            for (String messageId : messageIds) {
                messageIdMappings.remove(messageId);
            }
        }

        onMessagesChange();
        save();
    }

    public void markAsRead(String messageId) {
        Log.d(TAG, "setAsRead(" + messageId + ")");

        synchronized (this) {
            try {
                final DataDef.Message oldItem = messageIdMappings.get(messageId);
                if (oldItem == null) return;

                final int index = messages.indexOf(oldItem);
                final DataDef.Message newItem = oldItem.clone();
                newItem.setAsRead();

                messages.set(index, newItem);
                messageIdMappings.put(messageId, newItem);
            } catch (CloneNotSupportedException e) {
                e.printStackTrace();
            }
        }

        onMessagesChange();
        save();
    }

    public void markAllAsRead() {
        Log.d(TAG, "setAllAsRead()");

        synchronized (this) {
            for (int i = 0; i < messages.size(); ++i) {
                final DataDef.Message message = messages.get(i);

                try {
                    final DataDef.Message newItem = message.clone();
                    newItem.setAsRead();

                    messages.set(i, newItem);
                    messageIdMappings.put(message.getId(), newItem);
                } catch (CloneNotSupportedException e) {
                    e.printStackTrace();
                }
            }
        }

        onMessagesChange();
        save();
    }

    public DataDef.Message updateDeliveryInfo(String messageId, long mAlimtalkSendTime, long mMunjaSendTime, int mMunjaType) {
        Log.d(TAG, "updateDeliveryInfo(" + messageId + ", mAlimtalkSendTime: " + mAlimtalkSendTime + ", mMunjaSendTime: " + mMunjaSendTime + ", mMunjaType: " + mMunjaType + ")");

        final DataDef.Message message = messageIdMappings.get(messageId);
        if (message == null) return null;

        final int index = messages.indexOf(message);

        if (mAlimtalkSendTime > 0) {
            message.setAlimTalkSentAt(new Date(mAlimtalkSendTime));
        }

        if (mMunjaSendTime > 0) {
            message.setMunjaSentAt(new Date(mMunjaSendTime));
        }

        DataDef.MunjaType munjaType = DataDef.MunjaType.valueOf(mMunjaType);

        if (munjaType != null) {
            message.setMunjaType(munjaType);
        }

        messages.set(index, message);

        save();
        return message;
    }

    public synchronized void save() {
        Log.d(TAG, "save()");
        setValue(messages.toArray(new DataDef.Message[0]));
    }

    @Override
    public boolean delete() {
        instance = null;
        return super.delete();
    }

    // Private members
    private void loadRecentMessages() {
        final DataDef.Message[] array = getValue();
        if (array == null) return;

        final Date now = new Date();
        final int retainDay = Config.getInstance().getUmsMsgRetainDay();

        for (final DataDef.Message message : array) {
            long milliseconds = now.getTime() - message.getSentAt().getTime();
            long days = TimeUnit.DAYS.convert(milliseconds, TimeUnit.MILLISECONDS);


            if (days < retainDay) {
                messages.add(message);
                messageIdMappings.put(message.getId(), message);
            }
        }

        messages.sort((m1, m2) -> m2.getSentAt().compareTo(m1.getSentAt()));

        if (array.length != messages.size()) {
            save();
        }
    }

    private synchronized void onMessagesChange() {
        if (onMessagesChangeListener != null) {
            onMessagesChangeListener.onMessagesChange(messages);
        }
    }
}
